Data/ByteString/Lazy/ReadInt.hs

```diff
@@ -0,0 +1,286 @@
+readWord8 :: ByteString -> Maybe (Word8, ByteString)
+{-# NOINLINE readWord8 #-}
+readWord8 bs = _readWord q 5 bs
+  where q = fromIntegral @Word8 @Word (maxBound `div` 10)
+
+#else
+type Accum = Word64
+
+-- | A variant of 'readInt' specialised to 'Int64'.
+readInt64 :: ByteString -> Maybe (Int64, ByteString)
+{-# NOINLINE readInt64 #-}
+readInt64 bs = _readInt q bs
+  where q = fromIntegral @Int64 @Word64 (maxBound `div` 10)
+
+-- | A variant of 'readWord' specialised to 'Word64'.
+readWord64 :: ByteString -> Maybe (Word64, ByteString)
+{-# NOINLINE readWord64 #-}
+readWord64 bs = _readWord q 5 bs
+  where q = maxBound `div` 10
+#endif
+
+----- Handle optional signs and dispatch to the word-reader
+
+_readInt :: forall a. (Integral a, Bounded a)
+         => Accum -> ByteString  -> Maybe (a, ByteString)
+{-# INLINE _readInt #-}
+_readInt q bs = params >>= uncurry (_readWord q)
+  where
+    params = uncons bs >>= \ case
+        (w, s)
+              -- See above [Note @maxBound Int* `quotRem` 10 == (q, 7)@]
+            | fromDigit w <= 9 -> return (7, bs) -- leading digit
+            | w == 0x2d        -> return (8, s)  -- minus sign
+            | w == 0x2b        -> return (7, s)  -- plus sign
+        _                      -> Nothing        -- not a number
+
+----- ByteString layer unsigned read
+
+data Result = Overflow | Result !Int !Accum
+
+_readWord :: forall a. (Integral a, Bounded a)
+          => Accum -> Accum -> ByteString -> Maybe (a, ByteString)
+{-# INLINE _readWord #-}
+_readWord !q !r =
+#ifdef BYTESTRING_STRICT
+    \s@(BI.BS _ len) -> case _digits q r s 0 of
+        Result used acc
+            | used > 0
+              -> convert acc $!
+                 if used < len then BU.unsafeDrop used s else empty
+        _     -> Nothing
+  where
+#else
+    \ !s -> consume s 0 0
+    -- The second argument of consume, @valid@, tracks whether
+    -- we've parsed at least one decimal digit.  A numeric '0'
+    -- (rather than boolean 'False') is used so that updating the
+    -- value via @'(.|.)'@ does not create case branches that
+    -- would duplicate the inlined '_digits'.
+    --
+    -- We could use @'(+)'@, but at least in principle, if not in
+    -- practice, the addition could wrap around after @2^32@ bytes
+    -- of leading zeros on a 32-bit system!  Chunk lengths are
+    -- Ints, while lazy bytestring lengths are Int64s.  With
+    -- @'(.|.)'@ we get to /unask/ any questions about overflow.
+    --
+  where
+    consume :: ByteString -> Int -> Accum -> Maybe (a, ByteString)
+    -- All done
+    consume Empty !valid !acc
+        = if valid /= 0 then convert acc Empty else Nothing
+    -- Process next chunk
+    consume (Chunk c@(BI.BS _ !len) cs) !valid !acc
+        = case _digits q r c acc of
+            Result used acc'
+                | used == len
+                  -- process remaining chunks
+                  -> consume cs (valid .|. used) acc'
+                | (valid .|. used) /= 0
```

Comment:
The bitwise or doesn't get shared between the branches, resulting in a small amount of avoidable duplicated code. (likewise in ReadNat)